opt: Separate data driven testing framework from tests

andy-kimball opened the pull request:
The current opt test has two parts to it:
  1. Data-driven testing framework that dynamically loads the test
     case and expected result from a data file.
  2. The actual tests.

Already, the tests use multiple different kinds of commands, each
with its own input and result formats. Fast-forward a few years and
this code will become even more complex. This change separates the
parts so that #1 can be reused elsewhere. Even #2 should eventually
be broken up into separate files, one for each major chunk of
functionality (exec, normalization, build, etc). That will keep our
tests more decoupled from one another and easier to maintain.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21511)
<!-- Reviewable:end -->

andy-kimball committed:
opt: Separate data driven testing framework from tests

The current opt test has two parts to it:
  1. Data-driven testing framework that dynamically loads the test
     case and expected result from a data file.
  2. The actual tests.

Already, the tests use multiple different kinds of commands, each
with its own input and result formats. Fast-forward a few years and
this code will become even more complex. This change separates the
parts so that #1 can be reused elsewhere. Even #2 should eventually
be broken up into separate files, one for each major chunk of
functionality (exec, normalization, build, etc). That will keep our
tests more decoupled from one another and easier to maintain.

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> cc @RaduBerinde

I'm looking forward to using this to test the builder code!

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/opt/testutils/data_driven.go, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21511#-L39PHDfn5OZlKCeqsPe:-L39PHDfn5OZlKCeqsPf:b-453ii2) ([raw file](https://github.com/cockroachdb/cockroach/blob/834cc4fd1794fd6da641cc089ec4b1d209ac8644/pkg/sql/opt/testutils/data_driven.go#L1)):*
> ```Go
> // Copyright 2017 The Cockroach Authors.
> ```

[nit] 2018!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21511#-:-L39PDHH1HtEVGS8jcVt:bgg19tz)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/opt/testutils/data_driven.go, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21511#-L39PHDfn5OZlKCeqsPe:-L39cu2zW-LgPSXKoYm3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/834cc4fd1794fd6da641cc089ec4b1d209ac8644/pkg/sql/opt/testutils/data_driven.go#L1)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] 2018!
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21511)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21511)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Nice change, thanks!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21511#-:-L3BMKF8MszXf4nxuudv:brie5rq)*
<!-- Sent from Reviewable.io -->
